Setup.hs

```diff
@@ -1,18 +1,123 @@
-{-
-Copyright 2014 Google Inc. All Rights Reserved.
+{-|
+Module      : Main
+Description : Setup script for clang-pure build
+Copyright   : (C) Richard Cook, 2016
+Licence     : Apache 2.0
+Maintainer  : rcook@rcook.org
+Stability   : experimental
+Portability : portable
 
-Licensed under the Apache License, Version 2.0 (the "License");
-you may not use this file except in compliance with the License.
-You may obtain a copy of the License at
+This setup script adds a configuration hook to the build process to discover
+the LLVM library and include paths from the standard system-wide installation
+location or from an alternative location if overridden with the
+@CLANG_PURE_LLVM_DIR@ environment variable.
+-}
 
-    http://www.apache.org/licenses/LICENSE-2.0
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# OPTIONS_GHC -Wall #-}
 
-Unless required by applicable law or agreed to in writing, software
-distributed under the License is distributed on an "AS IS" BASIS,
-WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-See the License for the specific language governing permissions and
-limitations under the License.
--}
+module Main (main) where
 
+import Control.Exception
+import Control.Monad
+#ifndef mingw32_HOST_OS
+import Data.Maybe
+#endif
+import Distribution.PackageDescription
 import Distribution.Simple
-main = defaultMain
\ No newline at end of file
+import Distribution.Simple.LocalBuildInfo
+import Distribution.Simple.Setup
+import System.Directory
+import System.Environment
+import System.FilePath
+import System.IO.Error
+
+data SetupException = SetupException String deriving Show
+
+instance Exception SetupException
+
+data LLVMPathInfo = LLVMPathInfo
+    { llvmLibraryDir :: FilePath
+    , llvmIncludeDir :: FilePath
+    }
+
+llvmDirEnvVarName :: String
+llvmDirEnvVarName = "CLANG_PURE_LLVM_DIR"
+
+libraryExtension :: String
+#ifdef mingw32_HOST_OS
+libraryExtension = ".dll"
+#elif darwin_HOST_OS
+libraryExtension = ".dylib"
+#else
+libraryExtension = ".so"
+#endif
+
+getRecursiveContents :: FilePath -> IO [FilePath]
+getRecursiveContents dir = do
+    allNames <- getDirectoryContents dir `catch` (\e -> if isPermissionError e then return [] else throw e)
```

Comment:
`getDirectoryContents` also contains `.` and `..`, so this recurses forever. Also, this seems like a very expensive way to find libclang... I think it would be better to check the usual suspect locations (/usr/(local/)?lib/llvm-\d+(.\d+)*) and for other cases let the user set the environment variable.